README
NAME
    Apache::Logmonster - log utility for merging, sorting, and processing
    web logs

VERSION
    This documentation refers to Apache::Logmonster version 3.00

SYNOPSIS
    logmonster.pl -i <interval> [-v] [-r] [-n] [-b N]

       Interval is one of:

           hour    (last hour)
           day     (yesterday)
           month   (last month)

       Optional:

          -v     verbose     - lots of status messages 
          -n     dry run     - do everything except feed the logs into the processor
          -r     report      - last periods hit counts
          -b N   back N days - use with -i day to process logs older than one day

USAGE
    To see what it will do without actually doing anything

       /usr/local/sbin/logmonster -i day -v -n

    From cron:

       5 1 * * * /usr/local/sbin/logmonster -i day

    From cron with a report of activity:

       5 1 * * * /usr/local/sbin/logmonster -i day -v

DESCRIPTION
    Logmonster is a tool to collect log files from one or many Apache web
    servers, split them based on the virtual host they were served for, sort
    the logs into cronological order, and finally pipe the sorted logs to
    the log file analyzer of choice (webalizer, http-analyze, AWstats, etc).

  MOTIVATION
    Log collection: I have a number of web sites that are mirrored on two or
    three web servers. The statistics I care about are agreggate. I want to
    know how much traffic a web site is getting across all the servers. To
    accomplish that, the logs must be collected from each server.

    Sorting: Since most log processors require the log file entries to be in
    chronological order, simply concatenating them, or feeding them one
    after another does not work. Logmonster takes care of this by sorting
    all the log entries for each vhost into chronological order.

    Processor Agnostic: If I want to switch from one log processor to
    another, it should be simple and painless. Logmonster takes care of all
    the dirty work to make the possible. Each domain can even have its own
    processor.

  FEATURES
    * Log Retrieval from one or mnay hosts
    * Ouputs to webalizer, http-analyze, and AWstats.
    * Automatic vhost configuration
        Logmonster reads your Apache config files to learn about your
        virtual hosts and their file system location. Logmonster also
        generates config files as required (ie, awstats.example.com.conf).

    * Settings configuration for each virtualhost
        Outputs stats into each virtual domains stats dir, if that directory
        exists. This is an easy way to enable or disable stats for a virtual
        host. If "stats" exists, it will be updated. Otherwise it will not.
        Can also creates missing stats directories if desired (see
        statsdir_policy in logmonster.conf).

    * Efficient
        uses Compress::Zlib to read directly from .gz files to minimizes
        network and disk usage. Skips processing logs for vhosts with no
        $statsdir. Skips sorting if you only have logs from one host.

    * Flexible update intervals
        you can run it monthly, daily, or hourly

    * Reporting
        saves an activity report and sends an email friendly report.

    * Reliable
        lots of error checking so if something goes wrong, it gives a useful
        error message.

    * Apache savvy
        Understands and correctly deals with server aliases

INSTALLATION
    Step 1 - Download and install (it's FREE!)
	from CPAN or
        http://tnpi.net/cart/index.php?crn=210&rn=385&action=show_detail

        Install it The Perl Way:

           perl Makefile.PL
           make test
           make install 

	To install any missing dependencies, use 'make deps'

        To install the config file, use 'make conf' or 'make newconf'.
        newconf will overwrite any existing config file, so use it only for
        new installs.

    Step 2 - Edit logmonster.conf
         vi /usr/local/etc/logmonster.conf

    Step 3 - Edit httpd.conf
        Adjust the CustomLog and ErrorLog definitions. We make two changes,
        appending %v (the vhost name) to the CustomLog and adding cronolog
        to automatically rotate the log files.

          LogFormat "%h %l %u %t \"%r\" %>s %b \"%{Referer}i\" \"%{User-Agent}i\" %v" combined
          CustomLog "| /usr/local/sbin/cronolog /var/log/apache/%Y/%m/%d/access.log" combined
          ErrorLog "| /usr/local/sbin/cronolog /var/log/apache/%Y/%m/%d/error.log"

    Step 4 - Test manually, then add to cron.
          crontab -u root -e
          5 1 * * * /usr/local/sbin/logmonster -i day

    Step 5 - Read the FAQ
        <http://tnpi.net/wiki/Logmonster_FAQ>

    Step 6 - Enjoy
        Allow Logmonster to make your life easier by handling your log
        processing. Enjoy the daily summary emails, and then express your
        gratitude by making a small donation to support future development
        efforts.

DIAGNOSTICS
    Run in verbose mode (-v) to see addition status and error messages. If
    you want more verbosity, you can increase the verbosity by appending
    another -v (-v -v). If that is not enough, run with (-v -v -v) for even
    more verbosity. If that is not enough is, the source with you be. "Fear
    is the path to the dark side. Fear leads to anger. Anger leads to hate.
    Hate leads to suffering." Fear not, for the source is strong with you.

    Also helpful when troubleshooting is the ability to skip cleanup (so
    logfiles do not have to be fetched anew) with the --noclean command line
    option.

DEPENDENCIES
    Not perl builtins

      Compress::Zlib
      Date::Parse (TimeDate)
      Params::Validate

    Builtins

      Carp
      Cwd
      FileHandle
      File::Basename
      File::Copy

BUGS AND LIMITATIONS
    None known. Report problems to author. Patches are welcome.

AUTHOR
    Matt Simerson (matt@tnpi.net)

ACKNOWLEDGEMENTS
     Gernot Hueber - sumitted the daily userlogs feature
     Lewis Bergman - funded authoring of several features
     Raymond Dijkxhoorn - suggested not sorting the files for one log host
     Earl Ruby  - a better regexp for apache log date parsing

TODO
    Add support for analog.

    Add support for individual webalizer.conf file for each domain (this
    will likely not happen until someone submits a diff or pays me to do it
    as I don't use webalizer any longer).

    Delete log files older than X days/month - super low priority, it's easy
    and low maintenance to manually delete a few months log files when I'm
    sure I don't need them any longer.

    Do something with error logs (other than just compress)

    If files to process are larger than 10MB, find a nicer way to sort them
    rather than reading them all into a hash. Now I create two hashes, one
    with data and one with dates. I sort the date hash, and using those
    sorted hash keys, output the data hash to a sorted file. This is
    necessary as wusage and http-analyze require logs to be fed in
    chronological order. Look at awstats logresolvemerge as a possibility.

    Add config file setting for the location of awstats.pl

SEE ALSO
    http://tnpi.net/wiki/Logmonster

LICENCE AND COPYRIGHT
    Copyright (c) 2003-2007, The Network People, Inc. (info@tnpi.net) All
    rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are
    met:

    Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer.

    Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.

    Neither the name of the The Network People, Inc. nor the names of its
    contributors may be used to endorse or promote products derived from
    this software without specific prior written permission.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
    IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
    TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
    PARTICULAR PURPOSE ARE DIS CLAIMED. IN NO EVENT SHALL THE COPYRIGHT
    OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
    TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
    PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
    LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
    NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Logmonster ChangeLog

3.12 - Apr 2013
    testing changes, to resolve test errors on Windows

3.11 - Apr 2013
    testing & packing changes

3.09 - Apr 2013
    packaging changes

3.08 - Dec 2012
    added missing URL dependency
    updated bin/install_deps.pl

3.07 - Oct 2012
    removed Perl.pm
    updated Utility.pm to v 5.35
    stripped out apache httpd.conf parsing. Instead, detect vhosts from the log files (works with any http server)
    added confdir (path to processor config files)
    statsdir is now a full path to a directory
    the 'drop logs into vhost docroot" feature is gone, pending a clever way to reliably deduce the vhost document root for various web servers.
    removed vhost config option
    added default_vhost option
    moved _progress subs into Logmonster.pm
    removed get_domains_list, check_stats_dir, get_domains_list_from_directories, get_vhosts_from_file, install_default_awstats_conf, turn_domains_into_sort_key
    abstracted new subs: get_log_files, spam_check, open_vhost_handle, report_matches, report_spam_hits, report_bad_hits,
    added lighttpd config entries to instructions

3.05 - Sep 2008

    removed inc/* from distribution
    updated Makefile.PL, adding LICENSE and missing dependencies
    moved logmonster.pl into bin dir
    replaced bin/install_freebsd_deps.sh with install_deps.pl
    added 'make deps' target to Makefile.PL

3.04 - Nov 2007

    included "inc"' in modules
    Utility.pm - only attempt to use sudo as a last resort
    Utility.pm & Utility.t- synced with Mail::Toaster
    moved Regexp::Log to inc dir

3.03 - Feb 17, 2007

    Utility.pm - only attempt to use sudo as a last resort
    bin/install_freebsd_deps.sh - install Regexp::Log port

3.02 - Feb 16, 2007

    added support for detecting dynamic mass virtual hosting
    evaluate log for referrer spam and discard the referral spam.
    optional referral spam reporting
    the hits-per-vhost was only counting the number of hits for the master domain. Domain alias hits weren't showing up in the hits-per-vhost report.
    added bin/install_freeebsd_deps.sh installer
    default log location is now /var/log/http (was /var/log/apache)
    added Regexp::Log::Logmonster - uses a compiled regexp to parse log entries (very useful for evaluate for referrer spam).

  3.01  - Feb 2, 2007
    Added Params::Validate dependency to Makefile.PL

    Fixed bug where time offset was being ignored.

    Checks more locations for awstats.pl (needs to become a config file
    settings).

    Removed an unnecessary dependency on Mail::Toaster in Utility.pm.

  3.00rc2 - Oct 14, 2006
    *   updated usage examples

    *   added FAQ

    *   updated META.yml to new format

    *   email report formatting

        added OUTPUT_AUTOFLUSH b/c printing to *STDERR flushes immediately
        and mixed print statements did not. Email report is better formatted
        now.

    *   added inc/Module/Install

    *   added Build.PL

    *   added examples/daily_email_report.txt

    *   Logmonster.pm

        debug was not being set properly, report formatting tweaks

    *   Utility.pm

            added _progress sub

  3.00rc1 - Oct 02, 2006 (no public release)
   massive code overhaul
    This new version is mostly about code quality and maintainability (not
    new features). The large chunks of code have been modularized into
    smaller subroutines and tests have been written to test the
    functionality of each sub. There are now 102 different tests (was 23) in
    the test suite. Added t/Test-coverage.pl, t/Logmonster.t, t/pod.t,
    t/pod-coverage, t/00.load

    Nearly all the "working" code has been moved into
    lib/Apache/Logmonster.pm. Logmonster.pl is now a "shell" consisting of a
    little bit of code and a lot of documentation.

    All the functions are now Object Oriented. Time will tell if that is a A
    Good Thing[TM] but it makes reading the code and understanding where all
    the calls are going much, much easier. added doc/*

    The documentation has been significantly updated, addressing many of the
    common questions and comments I have received.

    The reporting has been overhauled. You still get the same information
    but by default, if everything is okay it runs entirely silently. A
    single -v will output status messages that make for a nice birds eye
    view of your web log traffic. You can add additional -v options for even
    more verbose reporting.

    Interface change: instead of -m for month, -d for day, and -h for hour,
    you use a -i [hour|day|month] option. The old -mdh options are
    officially deprecated but will continue to work for the indefinte
    future.

    synced lib/Apache/Logmonster/Perl & Utility with Mail::Toaster 5
    versions.

  2.77 - no public release
    *   create daily userlogs

        new feature submitted by Gernot Hueber.

    *   autogenerate vhost stat dirs

        new feature funded by Lewis Bergman: statsdir can be automatically
        created if it does not exist. Set statsdir_policy = create in
        logmonster.conf to enable this feature.

    *   incremental awstats config files.

        New feature by Lewis: Instead of having a full awstats.conf file for
        each vhost, each vhost has its own file that contains only specific
        info for that vhost and includes the generic
        /etc/awstats/awstats.conf file with the "global" settings.

    *   added loguser and loggroup to .conf

  2.76 - Jun 04, 2005
    statsdir can now be an absolute path
        statsdir can now be an absolute path (ex: /var/www/html). If so,
        processor output will be stored in statsdir/vhost. Otherwise, it
        works as it used to and output goes to vhost/statsdir. This is
        useful if you have a seperate machine (not the web server) that does
        the processing and that system does not have access to the vhost
        docroot.

  2.75 - Feb 05, 2005

Makefile.PL
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

use lib 'lib';

sub MY::postamble 
{

'LOGMONSTER = /Users/Shared/Sites/www.tnpi.net/internet/www/logmonster

docs:
	pod2text bin/logmonster.pl         > README
	pod2text doc/FAQ.pod               > FAQ
	pod2text doc/Changes.pod           > Changes
 
web:
	install -m 0644 bin/logmonster.pl $(LOGMONSTER)/logmonster.pl
	install -m 0644 logmonster.conf   $(LOGMONSTER)/logmonster.conf
	install -m 0644 doc/Changes.html  $(LOGMONSTER)/changes.html
	install -m 0644 doc/FAQ.html      $(LOGMONSTER)/faq.html
	install -m 0644 doc/README.html   $(LOGMONSTER)/README.html
	cp Apache-Logmonster-$(VERSION).tar.gz $(LOGMONSTER)/Logmonster.tar.gz
	mv Apache-Logmonster-$(VERSION).tar.gz $(LOGMONSTER)/Apache-Logmonster.tar.gz

conf:
	install -d /usr/local
	install -d /usr/local/etc
	install -m 0644 logmonster.conf /usr/local/etc/logmonster.conf-dist

newconf:
	install -d /usr/local
	install -d /usr/local/etc
	install -m 0644 logmonster.conf /usr/local/etc/logmonster.conf

deps:
	perl bin/install_deps.pl

';

};

WriteMakefile(
	'NAME'	           => 'Apache::Logmonster',
	'AUTHOR'           => 'Matt Simerson (msimerson@cpan.org)',
	'ABSTRACT'         => 'Apache log utility for merging, sorting, and processing web logs',
	'EXE_FILES'        => [ 'bin/logmonster.pl' ],
	'INSTALLSCRIPT'    => '/usr/local/sbin',
	'LICENSE'	       => 'bsd',
    'MIN_PERL_VERSION' => 5.008001,
	'VERSION_FROM'     => 'lib/Apache/Logmonster.pm',
	'PREREQ_PM'        => {
                            'Params::Validate'   => .8,
                            'Compress::Zlib'     => 2, 
                            'Date::Parse'        => 2,
                            'Regexp::Log'        => .04,
                            'Mail::Send'         => 0,
                            'URI'                => 1,
                            'Proc::ProcessTable' => 1, # Utility::is_process_running
                            'libwww-perl'        => 1,
                          },
    'TEST_REQUIRES'    => {
                          'Test::Pod'          => 0,
                          'Test::Pod::Coverage'=> 0,
                          },
    'BUILD_REQUIRES'   => {
                          'Test::Pod'          => 0,
                          'Test::Pod::Coverage'=> 0,
                          },
	 clean             => {
                            FILES => [  "t/trash",
                                        "t/tmp_*",
                                     ],
                          },
);



lib/Regexp/Log/Monster.pm
package Regexp::Log::Monster;

use strict;
use warnings;

use base qw( Regexp::Log );
use vars qw( %DEFAULT %FORMAT %REGEXP );

our $VERSION = 0.05;

=head1 NAME

Regexp::Log::Monster - A regexp parser for the Extended Log Format + vhost

=head1 SYNOPSIS

    my $foo = Regexp::Log::Monster->new(
        format  => ':logmonster';
        capture => [qw( ts request )],
    );

    # the format() and capture() methods can be used to set or get
    $foo->format('custom %date %request %status %bytes');
    $foo->capture(qw( ts req ));

    # this is necessary to know in which order
    # we will receive the captured fields from the regexp
    my @fields = $foo->capture;

    # the all-powerful capturing regexp :-)
    my $re = $foo->regexp;

    while (<>) {
        my %data;
        @data{@fields} = /$re/;    # no need for /o, it's a compiled regexp

        # now munge the fields
        ...
    }

=head1 DESCRIPTION

Regexp::Log::Monster uses Regexp::Log as a base class, to generate regular
expressions for performing the usual data munging tasks on log files that
cannot be simply split().

This specific module enables the computation of regular expressions for
parsing the log files created using the Monster Log Format. An example of
this format are the logs generated by the httpd web server using the
keyword 'common'.

The module also allows for the use of the Extended Monster Log Format.

For more information on how to use this module, please see Regexp::Log.

=head1 ABSTRACT

Regexp::Log::Monster enables simple parsing of log files created using the
Extended Log Format, such as the logs generated by the httpd web server 
using the keyword 'common'.

=cut

# default values
%DEFAULT = (
	format  => '%host %rfc %authuser %date %request %status %bytes %referer %useragent %vhost',
	capture => [ 'host', 'rfc', 'authuser', 'date', 'ts', 'request', 'req',
				 'status', 'bytes', 'referer', 'ref', 'useragent', 'ua', 'vhost' ],
);

# predefined format strings
%FORMAT = (
	':default'    => '%host %rfc %authuser %date %request %status %bytes %referer %useragent %vhost',
	':common'     => '%host %rfc %authuser %date %request %status %bytes',
	':extended'   => '%host %rfc %authuser %date %request %status %bytes %referer %useragent',
	':logmonster' => '%host %rfc %authuser %date %request %status %bytes %referer %useragent %vhost',
);

# the regexps that match the various fields
%REGEXP = (
#	'%host' => '(?#=host)\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}(?#!host)', 	# IPv4 only
	'%host'     => '(?#=host)\S+(?#!host)',					# numeric or name of remote host
	'%rfc'      => '(?#=rfc).*?(?#!rfc)',					# rfc931
	'%authuser' => '(?#=authuser).*?(?#!authuser)',				# authuser
	'%date'     => '(?#=date)\[(?#=ts)\d{2}\/\w{3}\/\d{4}(?::\d{2}){3} [-+]\d{4}(?#!ts)\](?#!date)',
										# [date] (see note)
	'%request'  => '(?#=request)\"(?#=req).*?(?#!req)\"(?#!request)',	# "request"
	'%status'   => '(?#=status)\d+(?#!status)',				# status
	'%bytes'    => '(?#=bytes)-|\d+(?#!bytes)',				# bytes
	'%referer'  => '(?#=referer)\"(?#=ref).*?(?#!ref)\"(?#!referer)',	# "referer"
	'%useragent'=> '(?#=useragent)\"(?#=ua).*?(?#!ua)\"(?#!useragent)',	# "user_agent"
	'%vhost'    => '(?#=vhost)\S+(?#!vhost)',				# name of local vhost
);

# note: date is in the format [01/Jan/1997:13:07:21 -0600]

1;

=head1 SEE ALSO

  Regexp::Log

=cut



logmonster.pl
#!/usr/bin/perl
use strict;
use warnings;

package Apache::Logmonster;

use vars qw($VERSION); 
$VERSION  = '3.04';

use lib "inc";
use lib "lib";

use English;
use Getopt::Long;
use Pod::Usage;
use Apache::Logmonster 3;
use Apache::Logmonster::Utility 5; 

my %command_line_options = (
    'bump:i'     => \my $bump,       # an optional time offset
    'clean!'     => \my $clean,      # ability to override conf file
    'interval:s' => \my $interval,   # hour/day/month
    'hourly'     => \my $hourly,
    'daily'      => \my $daily,
    'monthly'    => \my $monthly,
    'n'          => \my $dry_run,    # just show what we would do
    'verbose+'   => \my $verbose,    # incremental -v options
    'report'     => \my $report_mode,
);
if ( ! GetOptions (%command_line_options) ) {
    pod2usage;
};

# a generic utility object that provides many useful functions
my $utility = Apache::Logmonster::Utility->new();
my $banner  = "\n\t\t Apache Log Monster $VERSION by Matt Simerson \n\n";
my $config  = $utility->parse_config( file =>"logmonster.conf", debug=>0 );

# allow CLI to override clean option in config file
   $config->{'clean'} = $clean if defined $clean;
   $config->{'time_offset'} = $bump if defined $bump;

my $logmonster = Apache::Logmonster->new($config,$verbose);

# if this is not enabled, our report formatting will be jumbled
$OUTPUT_AUTOFLUSH++;

if ( $verbose && ! $report_mode ) {
  $verbose == 1 ? print "verbose mode (1).\n"
: $verbose == 2 ? print "very verbose mode (2).\n"
: $verbose == 3 ? print "screaming at you (3).\n"
: warn "unknown verbosity\n";
};

print $banner if $verbose;

# run a few preliminary sanity tests
$logmonster->check_config();

# CLI backwards compatability with previous versions
$interval ||= $hourly  ? "hour"
            : $daily   ? "day"
            : $monthly ? "month"
            : q{};

my %valid_intervals = ( hour => 1, day => 1, month => 1 );

if ( ! defined $valid_intervals{$interval} ) {
    pod2usage;
};

# stuff a few settings into the $logmonster object so
# Apache::Logmonster functions can access them.

my @hosts = split(/ /, $config->{'hosts'});
my $host_count = @hosts;
$logmonster->{'host_count'} = $host_count;
$logmonster->{'rotation_interval'} = $interval;
$logmonster->{'dry_run'} = $dry_run || 0;

if ($report_mode) {
    # prints out the last intervals hit count and exit, useful for SNMP
    $logmonster->report_hits();
    exit 1;
};

# open a file to log our activities to
my $REPORT = $logmonster->report_open("Logmonster", $verbose);

# store the file handle in the $logmonster object for functions
$logmonster->{'report'} = $REPORT;

$utility->_progress($banner) if $verbose;
print $REPORT $banner;

# do the work
my $domains_ref = 
$logmonster->get_domains_list     ();
$logmonster->fetch_log_files      ();
$logmonster->split_logs_to_vhosts ($domains_ref);
$logmonster->check_stats_dir      ($domains_ref);
$logmonster->sort_vhost_logs      () if !$dry_run;
$logmonster->feed_the_machine     ($domains_ref);
$logmonster->report_close         ($REPORT);

exit 1;   # happy exit status

__END__


=head1 NAME

Apache::Logmonster - log utility for merging, sorting, and processing web logs


=head1 VERSION
 
This documentation refers to Apache::Logmonster version 3.00
 

=head1 SYNOPSIS

logmonster.pl -i <interval> [-v] [-r] [-n] [-b N]

   Interval is one of:

       hour    (last hour)
       day     (yesterday)
       month   (last month)

   Optional:

      -v     verbose     - lots of status messages 
      -n     dry run     - do everything except feed the logs into the processor
      -r     report      - last periods hit counts
      -b N   back N days - use with -i day to process logs older than one day


=head1 USAGE

To see what it will do without actually doing anything

   /usr/local/sbin/logmonster -i day -v -n

From cron: 

   5 1 * * * /usr/local/sbin/logmonster -i day

From cron with a report of activity: 

   5 1 * * * /usr/local/sbin/logmonster -i day -v


=head1 DESCRIPTION

Logmonster is a tool to collect log files from one or many Apache web servers, split them based on the virtual host they were served for, sort the logs into cronological order, and finally pipe the sorted logs to the log file analyzer of choice (webalizer, http-analyze, AWstats, etc).


=head2 MOTIVATION

Log collection: I have a number of web sites that are mirrored on two or three web servers. The statistics I care about are agreggate. I want to know how much traffic a web site is getting across all the servers. To accomplish that, the logs must be collected from each server. 

Sorting: Since most log processors require the log file entries to be in chronological order, simply concatenating them, or feeding them one after another does not work. Logmonster takes care of this by sorting all the log entries for each vhost into chronological order.

Processor Agnostic: If I want to switch from one log processor to another, it should be simple and painless. Logmonster takes care of all the dirty work to make the possible. Each domain can even have its own processor.


=head2 FEATURES

=over

=item * Log Retrieval from one or mnay hosts

=item * Ouputs to webalizer, http-analyze, and AWstats.

=item * Automatic vhost configuration 

Logmonster reads your Apache config files to learn about your virtual hosts and their file system location. Logmonster also generates config files as required (ie, awstats.example.com.conf).

=item * Settings configuration for each virtualhost

Outputs stats into each virtual domains stats dir, if that directory exists. This is an easy way to enable or disable stats for a virtual host. If "stats" exists, it will be updated. Otherwise it will not. Can also creates missing stats directories if desired (see statsdir_policy in logmonster.conf).

=item * Efficient

uses Compress::Zlib to read directly from .gz files to minimizes network and disk usage. Skips processing logs for vhosts with no $statsdir. Skips sorting if you only have logs from one host.

=item * Flexible update intervals

you can run it monthly, daily, or hourly

=item * Reporting

saves an activity report and sends an email friendly report.

=item * Reliable

lots of error checking so if something goes wrong, it gives a useful error message.

=item * Apache savvy

Understands and correctly deals with server aliases

=back


=head1 INSTALLATION

=over

=item Step 1 - Download and install (it's FREE!)

http://tnpi.net/cart/index.php?crn=210&rn=385&action=show_detail

Install like nearly every perl module: 

   perl Makefile.PL
   make test
   make install 

To install the config file use "make conf" or "make newconf". newconf will overwrite any existing config file, so use it only for new installs.

=item Step 2 - Edit logmonster.conf

 vi /usr/local/etc/logmonster.conf

=item Step 3 - Edit httpd.conf

Adjust the CustomLog and ErrorLog definitions. We make two changes, appending %v (the vhost name) to the CustomLog and adding cronolog to automatically rotate the log files.

  LogFormat "%h %l %u %t \"%r\" %>s %b \"%{Referer}i\" \"%{User-Agent}i\" %v" combined
  CustomLog "| /usr/local/sbin/cronolog /var/log/apache/%Y/%m/%d/access.log" combined
  ErrorLog "| /usr/local/sbin/cronolog /var/log/apache/%Y/%m/%d/error.log"

=item Step 4 - Test manually, then add to cron.

  crontab -u root -e
  5 1 * * * /usr/local/sbin/logmonster -i day

=item Step 5 - Read the FAQ

L<http://tnpi.net/wiki/Logmonster_FAQ>

=item Step  6 - Enjoy

Allow Logmonster to make your life easier by handling your log processing. Enjoy the daily summary emails, and then express your gratitude by making a small donation to support future development efforts.

=back


=head1 DIAGNOSTICS

Run in verbose mode (-v) to see addition status and error messages. If you want more verbosity, you can increase the verbosity by appending another -v (-v -v). If that is not enough, run with (-v -v -v) for even more verbosity. If that is not enough is, the source with you be. "Fear is the path to the dark side. Fear leads to anger. Anger leads to hate. Hate leads to suffering." Fear not, for the source is strong with you.

Also helpful when troubleshooting is the ability to skip cleanup (so logfiles do not have to be fetched anew) with the --noclean command line option.


=head1 DEPENDENCIES

Not perl builtins

  Compress::Zlib
  Date::Parse (TimeDate)
  Params::Validate

Builtins

  Carp
  Cwd
  FileHandle
  File::Basename
  File::Copy


=head1 BUGS AND LIMITATIONS

None known. Report problems to author. Patches are welcome.


=head1 AUTHOR
 
Matt Simerson  (matt@tnpi.net)
 

=head1 ACKNOWLEDGEMENTS

 Gernot Hueber - sumitted the daily userlogs feature
 Lewis Bergman - funded authoring of several features
 Raymond Dijkxhoorn - suggested not sorting the files for one log host
 Earl Ruby  - a better regexp for apache log date parsing


=head1 TODO

Add support for analog.

Add support for individual webalizer.conf file for each domain (this will likely not happen until someone submits a diff or pays me to do it as I don't use webalizer any longer).

Delete log files older than X days/month - super low priority, it's easy and low maintenance to manually delete a few months log files when I'm sure I don't need them any longer.

Do something with error logs (other than just compress)

If files to process are larger than 10MB, find a nicer way to sort them rather than reading them all into a hash. Now I create two hashes, one with data and one with dates. I sort the date hash, and using those sorted hash keys, output the data hash to a sorted file. This is necessary as wusage and http-analyze require logs to be fed in chronological order. Look at awstats logresolvemerge as a possibility.

Add config file setting for the location of awstats.pl


=head1 SEE ALSO

http://tnpi.net/wiki/Logmonster


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2003-2007, The Network People, Inc. (info@tnpi.net) All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

Neither the name of the The Network People, Inc. nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DIS CLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut
